numpy/__init__.py

```diff
@@ -182,9 +182,37 @@
     oldnumeric = 'removed'
     numarray = 'removed'
 
-    # We don't actually use this ourselves anymore, but I'm not 100% sure that
-    # no-one else in the world is using it (though I hope not)
-    from .testing import Tester
+    if sys.version_info[:2] >= (3, 7):
+        # Importing Tester requires importing all of UnitTest
+        # which is not a cheap import
+        # Since it is mainly used in test suits, we lazy import it here
+        # to save on the order of 10 ms of import time for most users
+        # importing Tester also had a side effect of adding the full
+        # `numpy.testing` namespace
+        def __getattr__(attr):
+            import importlib
+
+            if attr == 'testing':
+                return importlib.import_module("." + attr, __name__)
```

Comment:
```
import numpy.testing as testing
return testing
```
should work? Although I am not sure how much of that is a language feature vs. an implementation detail, so maybe this is safer. I suppose this way will also throw of PyInstaller.

Reply:
Hmmm, i probably need to add a new test that ensures that things fail
![image](https://user-images.githubusercontent.com/90008/62257571-016efb80-b3d4-11e9-8ee2-c90446c0b612.png)

(Sorry for the screenshot, my terminal has been acting up recently)

Reply:
Honestly, I am not able to parse what you just said, I realize the code you have in the screenshot cannot work.

Reply:
```python
$ git diff
diff --git a/doc/sphinxext b/doc/sphinxext
--- a/doc/sphinxext
+++ b/doc/sphinxext
@@ -1 +1 @@
-Subproject commit a482f66913c1079d7439770f0119b55376bb1b81
+Subproject commit a482f66913c1079d7439770f0119b55376bb1b81-dirty
diff --git a/numpy/__init__.py b/numpy/__init__.py
index 5fbaf9aa8..fc61eacdc 100644
--- a/numpy/__init__.py
+++ b/numpy/__init__.py
@@ -193,7 +193,10 @@ else:
             import importlib

             if attr == 'testing':
-                return importlib.import_module("." + attr, __name__)
+                # import numpy.testing as testing
+                from . import testing
+                return testing
+                # return importlib.import_module("." + attr, __name__)
             elif attr == 'Tester':
                 from .testing import Tester
                 return Tester
@@ -207,7 +210,7 @@ else:
             # in __all__???
             not_added_to_all = ['bool', 'int', 'float', 'complex',
                                 'object', 'str', 'unicode']
-            return __all__ + ['Tester'] + not_added_to_all
+            return __all__ + ['Tester', 'testing'] + not_added_to_all

     else:
         # We don't actually use this ourselves anymore, but I'm not 100% sure that
diff --git a/numpy/tests/test_public_api.py b/numpy/tests/test_public_api.py
index 807c98652..bb9d23a31 100644
--- a/numpy/tests/test_public_api.py
+++ b/numpy/tests/test_public_api.py
@@ -69,6 +69,12 @@ def test_numpy_namespace():
     assert bad_results == whitelist


+    # make sure we can actually access `np.testing`, while not exported
+    # as part of the public API, it was accessible.
+    # With the use of __getattr__ and __dir__, this isn't always true
+    testing = np.testing
+
+
 def test_numpy_linalg():
     bad_results = check_dir(np.linalg)
     assert bad_results == {}

(numpy) mark@mark-xps ~/g/numpy import_time_tester|+3…3⚑ 1
$ ipython
iPython 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.6.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from numpy import testing
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
<ipython-input-1-79c9199de11a> in <module>
----> 1 from numpy import testing

~/miniconda3/envs/numpy/lib/python3.7/importlib/_bootstrap.py in _handle_fromlist(module, fromlist, import_, recursive)

~/git/numpy/numpy/__init__.py in __getattr__(attr)
    195             if attr == 'testing':
    196                 # import numpy.testing as testing
--> 197                 from . import testing
    198                 return testing
    199                 # return importlib.import_module("." + attr, __name__)

... last 2 frames repeated, from the frame below ...

~/miniconda3/envs/numpy/lib/python3.7/importlib/_bootstrap.py in _handle_fromlist(module, fromlist, import_, recursive)

RecursionError: maximum recursion depth exceeded

In [2]:
Do you really want to exit ([y]/n)? y
(numpy) mark@mark-xps ~/g/numpy import_time_tester|+3…3⚑ 1
$ pytest numpy/tests/test_public_api.py::test_numpy_namespace
Test session starts (platform: linux, Python 3.7.3, pytest 5.0.1, pytest-sugar 0.9.2)
rootdir: /home/mark/git/numpy, inifile: pytest.ini
plugins: forked-1.0.2, xdist-1.29.0, sugar-0.9.2
collecting ...
 numpy/tests/test_public_api.py ✓                                                                 100% ██████████

Results (0.02s):
       1 passed
```

I'm stumped, I can't seem to recreate a "failer" test.

Reply:
^^^^


I want to add a test that recreates the maximum recursion error, but when I try to access `numpy.testing` from pytest, it seems to work....

Reply:
I made a big hacky test with `subprocess.Popen` .... not really proud of it, but I couldn't find an other way of recreating the failure.

Reply:
To confirm - `import numpy.testing` will not hit `__getattr__`

Reply:
> To confirm - import numpy.testing will not hit __getattr__

This is correct.